pyteal/ast/jsonref.py

```diff
@@ -0,0 +1,107 @@
+from typing import TYPE_CHECKING
+from enum import Enum
+
+from pyteal.types import TealType, require_type
+from pyteal.errors import verifyFieldVersion, verifyTealVersion
+from pyteal.ir import TealOp, Op, TealBlock
+from pyteal.ast.expr import Expr
+from pyteal.ast.leafexpr import LeafExpr
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class JsonRefType(Enum):
+    # fmt: off
+    #              id |    name   |      type      | min version
+    string = (0, "JSONString", TealType.bytes,  7)
+    uint64 = (1, "JSONUint64", TealType.uint64, 7)
+    object = (2, "JSONObject", TealType.bytes,  7)
+    # fmt: on
+
+    def __init__(self, id: int, name: str, type: TealType, min_version: int) -> None:
+        self.id = id
+        self.arg_name = name
+        self.ret_type = type
+        self.min_version = min_version
+
+    def type_of(self) -> TealType:
+        return self.ret_type
+
+
+JsonRefType.__module__ = "pyteal"
+
+
+class JsonRef(LeafExpr):
+    """An expression that accesses the value associated with a given key from a supported utf-8 encoded json object.
+
+    The json object must satisfy a [particular specification](https://github.com/algorand/go-algorand/blob/master/data/transactions/logic/jsonspec.md).
+    """
+
+    def __init__(self, type: JsonRefType, json_obj: Expr, key: Expr) -> None:
```

Comment:
@jdtzmn Minor and optional:  There's 2 places where a parameter is named `type`.  If you're on-board, I propose renaming because there's a builtin `type` available by default.  

The parameters shadow the builtin `type` definition.  In my experience, lexical shadowing is error prone.  

That said, I don't feel too strongly.  If you prefer as is, feel welcomed to resolve the thread.

Reply:
If it were the only location in the codebase which used `type` as a parameter, I'd feel differently. But since it appears in many other places I'll resolve the conversation instead.